### ISSUE TITLE
[BUG] Training hangs after step 0 — asymmetric NCCL collectives in finalize_model_grads

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#51]
### Summary

Training consistently hangs after completing step 0 when using pipeline parallelism (PP > 1) with gradient accumulation. The deadlock occurs inside `finalize_model_grads()` during the allreduce of embedding weight gradients.

### Environment

- **Neuron_SP version**: `main` @ commit `a3f9c12`
- **PyTorch**: 2.3.0+cu121
- **NCCL**: 2.20.5
- **CUDA**: 12.4
- **GPU**: 8x H100 SXM5
- **Config**: PP=4, TP=2, DP=1, `--use-distributed-optimizer`, gradient accumulation steps=4

### Reproduction Steps

```bash
torchrun --nproc_per_node=8 pretrain_gpt.py     --pipeline-model-parallel-size 4     --tensor-model-parallel-size 2     --micro-batch-size 2     --global-batch-size 32     --num-layers 32     --hidden-size 4096     --train-iters 100
```

Process hangs indefinitely after printing `[step 0] loss: 11.34...`

### Root Cause (Suspected)

In `megatron/training/utils.py::finalize_model_grads()`, the embedding allreduce is conditionally called only on ranks that own the embedding layer:

```python
if mpu.is_pipeline_first_stage(ignore_virtual=True):
    embedding_grads_all_reduce(model)
```

However, the NCCL communicator group `_EMBEDDING_GROUP` is initialized for **all** data-parallel ranks. When PP stages > 2, non-first-stage ranks enter a different allreduce path (for tied weights), creating an asymmetric collective — exactly one side of the communicator is never called, causing NCCL to deadlock.

### Expected Behavior

All ranks in `_EMBEDDING_GROUP` participate symmetrically in the same collective operation.

### Actual Behavior

- Ranks 0, 4 (first PP stage): call `ncclAllReduce` on `_EMBEDDING_GROUP`
- Ranks 1, 2, 3, 5, 6, 7: skip the call and proceed to optimizer step
- NCCL hangs waiting for missing participants

### Proposed Fix

Add an explicit `torch.distributed.barrier()` scoped to `_EMBEDDING_GROUP` before the conditional, or refactor to use `reduce` → `broadcast` pattern so all ranks participate:

```python
# Option A: barrier guard
dist.barrier(group=mpu.get_embedding_group())
if mpu.is_pipeline_first_stage(ignore_virtual=True):
    embedding_grads_all_reduce(model)

# Option B: symmetric reduce-broadcast (preferred)
dist.reduce(emb_grad, dst=first_stage_rank, group=mpu.get_embedding_group())
if mpu.is_pipeline_first_stage(ignore_virtual=True):
    # scale and process
    dist.broadcast(emb_grad, src=first_stage_rank, group=mpu.get_embedding_group())
```

### Priority

**P0** — blocks all PP > 1 training runs. No workaround available.
